### ISSUE TITLE
Unify `RenderLayers` and `TargetCamera` and provide a single way to configure which camera entities are rendered to

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#12468]
## What problem does this solve or what need does it fill?

`TargetCamera` (UI) and `RenderLayers` (not UI) both do the same thing: control which cameras objects are rendered to.

We should have one way, that works in a consistent manner across both.

## What solution would you like?

Just use a `HashSet<Entity>` of cameras in an extended `TargetCamera` component. This is simple, doesn't have an arbitrary limit and doesn't involve coordination. It can also be cleanly replaced with a custom relation.

## What alternative(s) have you considered?

Move everything over to `RenderLayers`. This may be faster: there's less lookup involved. It's also more traditional: this is how other engines have historically handled this.

## Additional context

A [new user was bitten by this today](https://discord.com/channels/691052431525675048/691052431974465548/1217672343111860295), and the behavior seems buggy, especially in mixed UI and non-UI trees. This idea came up independently in [`ui-dev` discussions](https://discord.com/channels/691052431525675048/1217481411175649340/1217516897411661925) with @aevyrie.

**Comment by tychedelia** [comment]
I ran into the [32 limit](https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/view/visibility/render_layers.rs#L5) for `RenderLayers` in an application I'm building. In my case, I need to sort entities across multiple windows as well as use many cameras to render individual entities into the ui.

I attempted to work around this by changing `RenderLayers` to be a growable bitset. Unfortunately, `wgpu` only allows dynamic arrays in the last position in a uniform, and I ran into issues implementing this as `RenderLayers` are not only used on the `View` uniform, but also `DirectionalLight`s, which are stored in an array themselves.

Currently, it seems that `RenderLayers` are only being used in our shaders for filtering out these lights.

In terms of my use case, the ability to associate entities with multiple cameras is unnecessary. I really just want 1:1 mappings of entities to cameras.

**Comment by tychedelia** [comment]
So it seems like a first step would be removing our use of `RenderLayers` from our shader code, which currently only happens in [one place](https://github.com/bevyengine/bevy/blob/main/crates/bevy_pbr/src/render/pbr_functions.wgsl#L272). The suggestion here would be to move lights to a view and compute the visible lights on the cpu (using `RenderLayers` or whatever alternative we develop).

We want to make sure that we maintain support for entities to target multiple cameras, which is currently supported by `RenderLayers`. This was brought up on discord, but certainly sounds like an important use case for relations.

**Comment by tychedelia** [comment]
> Unify the grouping logic:

While this would solve my particular problem, I'm not sure I love the api design. Besides the limit on their size, `Layers` are more powerful than `Camera(Entity)` and so it may not be clear to users why you'd want to use one over the other. Additionally, we have  the concept of "default" in two places here. And while it's not super complicating, this would add more complexity to every place we check whether an entity should be rendered.

I think we should try to generalize this API by removing the limit on `RenderLayers`, either through a growable bitset which allows an arbitrary number of layers of just by waiting until we have relations, which can better express M:N relations between entities and cameras.

**Comment by viridia** [comment]
See also #12461

Inheritability is a feature of `TargetCamera`, but not of `RenderLayers`; but I would very much like inheritability for non-UI entities. A prime use case is GLTF scenes: currently, when you spawn a `SceneBundle`, you can insert a `RenderLayers` component, but this has no effect on the meshes contained within the scene. Thus, if you want a given instance to appear in one layer only, or multiple instances of the same scene to appear in different places in different layers (which my game depends on greatly) you have to decompose the GLTF scene and mutate the nodes, a complex process.

So if we're looking to revamp layers, I'd lobby strongly for a solution that supports inheritance for both ui and non-ui nodes.

**Comment by alice-i-cecile** [comment]
Yep, we definitely shouldn't and don't need to block on relations.

**Comment by viridia** [comment]
BTW, I wanted to describe in more detail my particular use case for render layers. Let's say the player is traveling in the outdoor wilderness and comes across a cave entrance. The cave entrance is a portal, which leads to an underground cave complex. The portal connects two "realms", the outdoor realm and the underground realm, both of which exist within the same Bevy World, but which are separate "environments" or scenes. (I'm hesitant to use the word "scene" here because it implies that the environment is a unitary asset, when in reality the environment may be assembled algorithmically from multiple scenes.)

Render layers are used to keep the realms separate: the outdoors is render layer 1, which is rendered on the main camera; the cave interior is render layer 2, which is rendered via the portal camera to an offscreen render target and then composited into the outdoor scene. Once the player enters the cave, the primary camera will switch to layer 2, and the portal camera will switch to layer 1.

Now let's say we have some common GLTF object, such as a crate or treasure chest. We can place instances of the chest in the cave, and we can also place additional instances of the chest in the outside of the cave entrance. This means that the chest instances need to have render layer settings which match their environments. Having to place render layer components on each individual sub-mesh is cumbersome, especially as it requires unpacking assets and cloning their contents individually.

It would be much better for the user if they could simply spawn a SceneBundle and set the render layer once on the bundle - or even better, set the render layer on the root of the entire environment and have it apply to all descendants.

**Comment by tychedelia** [comment]
Sorry, to be clear, I was not suggesting that we *should* block on relations, just that this is a uniquely suited problem to that feature, and so I would be hesitant to overly complect something that could be greatly simplified later. `RenderLayers` are basically an ad hoc join table, and so would benefit from databasification.

> Visibility problem space

I appreciate this breakdown and think it helps clarify the problem space. My original objection was just that `RenderLayers` are equivalent to a `TargetCamera` where each camera has a unique layer, and hence a superset of that behavior.

In particular, I think it helps illustrate the difference between `RenderLayers` and @alice-i-cecile's suggestion of `HashSet<Entity>`, which is that that `RenderLayers` allow cameras to "subscribe" to groups of entities without entities knowing about that camera. In the case of `HashSet<Entity>`, spawning a new camera would require doing a query for those entities and inserting the new camera entity. In practice, I am not sure about use cases for this behavior, and I think providing some could help clarify this discussion, as it's possible this is a rather niche use case, in which case `HashSet<Entity>` would *greatly* simplify things and be useful as a replacement for both these existing patterns.

> What's not clear to me is how you'd do Camera : layer mappings. When you e.g. spawn a new UI camera, do you need to manually request a new bit from a central layer registry? Then what about cleanup if you are dynamically spawning/despawning cameras? What if there are TONS of cameras?

This is definitely a bit of bookkeeping and a drawback of `RenderLayers` for dynamic cameras. In my app, I am using a layer registry, which isn't ideal but also wasn't that hard to implement.

With further consideration, I think expressing this as `TargetCameras<HashSet<Entity>>` would be the best middle ground if we're okay with sacrificing the "subscription" behavior of `RenderLayers`.

**Comment by tychedelia** [comment]
> Once the player enters the cave, the primary camera will switch to layer 2, and the portal camera will switch to layer 1.

Well, this is a great example of the kind of behavior I was asking for examples of in my previous comment! :) I'm trying to think, but this seems more difficult to manage with `HashSet<Entity>` alone. Although, I guess you could instead swap the `RenderTarget` for each camera?

**Comment by tychedelia** [comment]
> An entity is Global and then World-group(1) is added. Does the entity stay Global unless that designation is manually removed?

I think @viridia's example points us in a direction here -- in this case, there is no "global" grouping. Rather, there is a logical "current" grouping that renders to the window now, but may be replaced by another group later. This is also the case for the multi-window application I'm developing in which entities are always association with a particular window-camera pairing. The primary window could be considered global, but really it's just another window-camera pair. This is why inheritability is important, as we want to typically make "broad strokes" groupings of entities. I think this also points towards a kind of pattern of feature here that is a bit broader than just camera targeting and maybe deserves more design exploration, i.e. scene groupings, nested worlds, etc.

In general, I think adding a grouping suggests that the entity is no longer global.

> An entity is Global and then Camera-affiliated(some_camera) is added. Does the entity stay Global? I don't think you should have Global for UI nodes (at least by default). 

Adding a camera target indicates something is not global.

> An entity is Camera-affiliated(some_camera) and then World-group(1) is added. Is the entity visible both to some_camera and to any camera subscribed to World-group(1)?

**Comment by viridia** [comment]
Well, except for the inheritability issue, I can make bitmasks work for my use case. It does mean some fancy footwork in managing environments: if I have more than 32 realms, then I need to dynamically assign the layer bits, especially if some layers are being dedicated for other purposes. But there's absolutely no way I am going to be rendering 32 portals in a single frame, that would be unacceptably slow. Yes, switching the layer bits dynamically might be expensive, but it generally only happens during a transition, and I have to cover over the asset (pre-)loading with smoke and mirrors anyway, that's what makes game programming an art and not a science :)

If render layers are inheritable, then target camera can simply be replaced by the same bitmask: have a dedicated bit (bit 31 say) for the default UI, and let the user define additional UI bits as needed.

One option is to simply use an int64 mask and double the number of available layers, with bit 63 being the default UI bit.

**Comment by viridia** [comment]
BTW, I should also point out that in my multi-environment example, it's important to get the render layer bits for the lighting correct. Each "realm" must have its own independent directional lights, because otherwise you get shadows from the various environments mixed together in the same shadow buffer. Given that shadow maps are expensive, that puts a constraint on how many active environments you have loaded in memory, completely independent of the number of available layer bits.

**Comment by tychedelia** [comment]
> The solution we choose should not be at most a small perf regression. Visibility is definitely on the hot path. I am concerned that HashSet is much more expensive than the current bitmask approach for lookups, for propagating visibility settings, and that's a lot of additional allocations for the base-layer of Bevy.

Yes, `HashSet` definitely won't work if we assume inheritability or a broad use case, this is why a growable bitset was my first option.

**Comment by viridia** [comment]
I'm find with a fixed-size set of bits so long as we can make a convincing case that the number of bits available is sufficient to cover the possible use cases. But this requires iterating over potential use cases, which is hard.

Let me extend my previous example for a moment: suppose that our game is in fact an online chat environment with avatars, with portals between realms that represent different servers. In such a case, the number of realms is for all practical purposes infinite. However, no game engine is going to be able to manage an infinite number of realms, or even a very large number; it's up to the engine to cull the available environments to a reasonable number based on context. I'd imagine that even a dozen portals is pushing the limits of user-comprehensible scene complexity. However, the number of portals and the number of realms are only loosely correlated: in my own engine, portals are only active when they are (a) within the current view frustum and (b) facing the camera, so a realm might be "nearby" (in the sense that the remote environment geometry is cached in memory) but not actually drawn at any given moment.

My point is that there's downward pressure coming from a number of factors: number of render layers is one, but number of cameras, lights, meshes, shadow buffers and so on all contribute to the equation.

**Comment by tychedelia** [comment]
> My point is that there's downward pressure coming from a number of factors: number of render layers is one, but number of cameras, lights, meshes, shadow buffers and so on all contribute to the equation.

Absolutely, and I've also said before that I could *probably* live with a fixed bitset, but 32 is too low. My use case is a bit more niche: I'm building a desktop app and need to render a large number of ~1-5 entity "scenes" to textures for use throughout the ui. However, sometimes these scenes need to both be rendered as a texture in the ui and on a separate window's surface. It's more important here to remove constraints to let users do whatever they want, with the assumption that they'll optimize if they run into problems.

**Comment by viridia** [comment]
For the inheritance feature, I was thinking that you'd have two components, similar to `Visibility` and `InheritedVisibility` (or `Transform`/`GlobalTransform`) - one of which is added by the user, the other maintained automatically. Since walking the tree is expensive, you could perform this computation at the same time as the visibility update.

**Comment by eidloi** [comment]
> The [proposed API](https://github.com/bevyengine/bevy/issues/12468#issuecomment-1998961486) probably does not have acceptable efficiency, because adding multiple cameras definitely requires allocating.
> 
> I think the camera-affiliation approach is very useful and ergonomic for UI, where you only need one camera for a big group of entities and shouldn't need to do any bit fiddling when adding/removing UI cameras. So a simplified API would allow only 1 camera entity, plus a growable bitset of layers.
> 
> ```rust
> struct RenderGroup
> {
>     layer0: bool,
>     layers: SmallVec<[u8; 4]>,
>     camera: Option<Entity>,
> }
> ```

Will this allow use cases like a minimap for instance? Where you have an overhead camera that renders _some_ parts of the terrain to a render texture with annotations on it? I mean this would need some entities to be rendered on  two cameras OR we need to duplicate said entities and target different cameras with them. This is an extreme example in the sense that minimaps usually render marker Entities that aren't visible on the main camera, but there will probably be situations where we don't want to duplicate entities just to render them twice. _Though it might be more efficient in the end_

**Comment by eidloi** [comment]
We don't need to rewrite the `InheritedRenderGroups` every frame if the system operates on change detection (so added, removed, changed `Children` of either a `PropagateRenderGroups` or an `InheritedRenderGroups`, with an additional check on the changed or removed `Parent` of the `InheritedRenderGroups` entities). It might be a little less performant when you only have a small set of entities, but larger trees should greatly benefit I think.

**Comment by eidloi** [comment]
Though in this case we are talking about potential chugs when a larger tree changes, VS a constant (steady) performance cost...

**Comment by eidloi** [comment]
It might also be interesting to short-circuit changes when (after a re-parent) the new parent has the same inherited render groups. Which means the children of the re-parented entity doesn't need to update at all

**Comment by tychedelia** [comment]
> Now I will try to sketch a full visibility API, renaming `RenderGroup` to `RenderGroups`.

This is making sense to me and I think I've been convinced that it's useful to have both layers and target camera. Thanks for spelling this all out.

However, I'm still a bit confused around the logic regarding `layer0` or what you refer to as "plain" entities. In the current visibility code, if an entity or. view is missing a `RenderLayers` component, we give it `RenderLayers::default()` aka layer 0. I'm not sure I understand what the purpose of the `layer0` field is in your sketch.

> A camera cannot see both plain and layer entities. This is because `!RenderLayers::layer(0).intersects(RenderLayers::layer(0).with(1))`. 
>
> ...
>
> I include the RenderGroup::Layers::default field so you can view 'plain' entities in addition to layer entities.

I'm a bit confused by this. A mask with layer 0 and layer 0+1 *do* intersect. In other words, if a camera has a mask of layer 0+1, it will see any "plain" entities that do not have a `RenderLayers` component.  If I give a camera *just* layer 1, I would not expect it to be able to see "plain" entities in that case. Similarly, if I give an entity layer 1, I would not expect a camera without a `RenderLayers` component to be able to see it.

Can you give an example of how this is intended to be used?

> It may be more performant to just use u32 in the RenderLayers SmallVec. Not sure how valuable byte-shaving is here.

Just a minor note, but we have [fixedbitset](https://docs.rs/fixedbitset/latest/fixedbitset/index.html) in our dependency tree which would probably be a better fit than `SmallVec`. It has a `grow` method that we can use to ad hoc increase the size of the render layers when a new layer would be greater than our current size.

**Comment by viridia** [comment]
I'm not sure fixedbitset is all that efficient in this case, a quick glance at the source code reveals that it stores the bits in a Vec, which means that if you use even 1 bit you are going to alloc memory.

**Comment by alice-i-cecile** [comment]
Write-up from @viridia about some of the use cases and challenges involved: https://gist.github.com/viridia/f761f1813a9edeca2ba90b1d8468160d
